src/main/rx/rx.c

```diff
@@ -1023,26 +1022,19 @@ bool isRssiConfigured(void)
     return rssiSource != RSSI_SOURCE_NONE;
 }
 
-timeDelta_t rxGetFrameDelta(timeDelta_t *frameAgeUs)
+timeDelta_t rxGetFrameDelta(void)
 {
-    // returns the delta time based on the Rx Driver time stamps
+    // return the delta time based on time stamps provided by the Rx driver
+    // frameTimeDeltaUs holds its last good value until after 150ms of no frames, it goes to zero)
     static timeUs_t previousFrameTimeUs = 0;
-    static timeDelta_t frameTimeDeltaUs = 0; // last valid Rx-reported frame time stamp delta
+    static timeDelta_t frameTimeDeltaUs = 0;
 
-    // when a new packet arrives, return the delta time 
-    if (rxRuntimeState.rcFrameTimeUsFn) {
-        const timeUs_t frameTimeUs = rxRuntimeState.rcFrameTimeUsFn(); // Frame time stamp from Rx driver
+    const timeUs_t frameTimeUs = rxRuntimeState.rcFrameTimeUsFn(); // Frame time stamp from Rx driver
+    DEBUG_SET(DEBUG_RX_TIMING, 1, frameTimeUs);     // time from reception of frame to now in hundredths of ms
 
-        *frameAgeUs = cmpTimeUs(micros(), frameTimeUs);
-        // Time interval between now and most recent Rx-reported frame time
-        // steps up every 100ms if Rx does not update the last reported frame time stamp
-
-        const timeDelta_t deltaUs = cmpTimeUs(frameTimeUs, previousFrameTimeUs);
-        if (deltaUs) {
-            // Rx-reported frame time has changed, compared to previous Rx-reported frame time stamp
-            frameTimeDeltaUs = deltaUs;
-            previousFrameTimeUs = frameTimeUs;
-        }
+    if (frameTimeUs) {
```

Comment:
`frameTimeUs` is not reset after reading it, so this will return nonzero `frameTimeDeltaUs` only once per 'rxRuntimeState.lastRcFrameTimeUs' update. All other calls will return 0.

Depending on the fact that `rxGetFrameDelta` is called only after 150ms is pretty bad idea ...

Reply:
To be honest, this part of the code confused me greatly.
I'm not sure if I'm answering your concerns in this response, if not I'd welcome further advice.
On logging, the value of `frameTimeDeltaUs` (orange line in image below) drops to zero at 150ms.
The total gap in the Rx signal in this dropout is about 300ms.  The 150ms initial lag is between the two red centre line markers, then the estimated delta 3 steps up at 50ms intervals.
Note that `frameTimeDeltaUs` is locked at its last value for 150ms, but then drops to zero until there is a new value.  I'm not sure why it goes to zero. 
This behaviour isn't so bad because the interval then drops back to the non-Rx-driver method, which steps up progressively as the packet loss period persists.
To be honest, I'm not sure why we take the effort to check the Rx-reported time stamps and get the rx-reported delta time.  There seems to be, certainly for CRSF, not much of a difference at all, in normal cases, between the simple delta time calculated as `frameDeltaUs` and the Rx reported value of `frameDeltaUsRx`.  The only big difference is with long dropouts, and there we nevert to the simpler `frameDeltaUs` value as an estimate of the duration, since `frameDeltaUsRx` goes to zero.

PS - why do we not use `rxRuntimeState->lastRcFrameTimeUs` in this case?

Reply:
> To be honest, this part of the code confused me greatly. 

That is why I'm trying to reduce ad-hoc/heuristic PRs. You can try following git blame, but IMO this is result of fixing problem in wrong place / hacking RX detection to work.

> On logging, the value of `frameTimeDeltaUs` (orange line in image below) drops to zero at 150ms. 
The total gap in the Rx signal in this dropout is about 300ms. The 150ms initial lag is between the two red centre line markers, then the estimated delta 3 steps up at 50ms intervals. Note that `frameTimeDeltaUs` is locked at its last value for 150ms, but then drops to zero until there is a new value. I'm not sure why it goes to zero. 

This is strange. It doesn't match neither old, nor new code.


> This behaviour isn't so bad because the interval then drops back to the non-Rx-driver method, which steps up progressively as the packet loss period persists. 

So there may be a problem, but it is carefully masked  ...


> To be honest, I'm not sure why we take the effort to check the Rx-reported time stamps and get the rx-reported delta time. There seems to be, certainly for CRSF, not much of a difference at all, in normal cases, between the simple delta time calculated as `frameDeltaUs` and the Rx reported value of `frameDeltaUsRx`. 

There is RC smoothing code that depends on accurate RC rate estimate (or, strictly speaking, results do not look nice in blackbox without it). 

> PS - why do we not use `rxRuntimeState->lastRcFrameTimeUs` in this case?

rx_spi is different.


Also note that rcFrameTimeUsFn may be NULL for some receivers (at minimum for rx_spi  without exti) . Old code did handle that case, this PR will hardfault.

Reply:
Thanks @ledvinap for your feedback!  I agree that this part of the Rx code is definitely strange.  I'll check git blame to try to figure out why it is the way it is.

> There is RC smoothing code that depends on accurate RC rate estimate...

Yes, and I did a lot of that coding, and am aware that these intervals are primarily used for RC Smoothing; they also are used for OSD link rate display, and the delta time in feedforward.  The min/max limits on `currentRxIntervalUs` 'hide' some of the errors in this area from those estimates.  Feedforward perhaps needs the full delay time, not the constrained time, but that's another topic.  I can look more at the 'application of the data' later.  The goal here, we agree, is to get this data as accurate as it can reasonably be, without unduly complex coding.

I will log and compare the time delta derived from the Rx driver time stamps in `rxRuntimeState->lastRcFrameTimeUs`, ie `frameDeltaUsRx`, to the simpler method used in `updateRcRefreshRate`.  

There may have been differences previously, but now that the scheduler checks for new Rx data every loop, there is no meaningful delay, I think, in getting packet intervals based on when `updateRcRefreshRate` runs.  There is probably now no difference from calculating delta times based on when the crsf.c driver code, for example, runs.  If so we don't really need to worry about using `rxRuntimeState->lastRcFrameTimeUs` at all.

However if we force `rxRuntimeState->lastRcFrameTimeUs` to zero when it is not usable (test for zero), and then revert to the `currentRxIntervalUs` value when it is, that should be OK in the interim, while we get some data.

> Also note that rcFrameTimeUsFn may be NULL for some receivers (at minimum for rx_spi without exti) . Old code did handle that case, this PR will hardfault.

How would I best solve that?

Reply:
Update... I put the original  check back, which should solve the NULL possibility, assuming the old code worked.
I logged `rxRuntimeState->lastRcFrameTimeUs` and it never went to zero with CRSF.  Makes for a nice way to see the overall steps.  Cyan sloping line; any flat spot = no new data.
The reason `frameDeltaUsRx` drops to zero is that the code calculating it is triggered at regular intervals during dropouts.
The calling function uses zero as an indicator that the Rx driver supplied value isn't useful, and then uses the `frameDeltaUs` value instead.
This works pretty well I think.
Here are two dropout episodes
![Screen Shot 2024-03-13 at 12 37 40](https://github.com/betaflight/betaflight/assets/11737748/9ad0a533-ebd8-4b37-a141-514828104c0d)

Reply:
> There may have been differences previously, but now that the scheduler checks for new Rx data every loop, there is no meaningful delay, I think, in getting packet intervals based on when updateRcRefreshRate runs. There is probably now no difference from calculating delta times based on when the crsf.c driver code, for example, runs. If so we don't really need to worry about using rxRuntimeState->lastRcFrameTimeUs at all.

There may be differences in other drivers. Protocols with non-uniform packet length will be one example. 
Also, RC processing is quite low priority, scheduling may be less predictable in future.
But readout virtual function is overkill (and it's implementation is wrong in rx_spi case - it may return timestamp of packet that was not fully received yet)

Reply:
> The reason frameDeltaUsRx drops to zero is that the code calculating it is triggered at regular intervals during dropouts.
The calling function uses zero as an indicator that the Rx driver supplied value isn't useful, and then uses the frameDeltaUs value instead.
This works pretty well I think.

Isn't it terribly overcomplicated?

Reply:
Less complicated than before :-) and yet I agree, still too complicated. 

Let me know what suggestions you have for simplifications. 

I don’t know anything about the SPI code and would prefer any changes to SPI or MSP data streams to be in a separate PR.

Reply:
ad SPI/MSP  - it should be quite trivial to update lastRcFrameTimeUs when packet is received.